### ISSUE TITLE
Conditional Types - Checking `extends never` only works sometimes

### ISSUE THREAD

**Issue by dsherret** [Microsoft/TypeScript#23182]
**TypeScript Version:** 2.9.0-dev.20180405

**Search Terms:** "extends never"

**Code**

I am writing a [function](https://github.com/dsherret/conditional-type-checks/blob/master/index.ts#L5) for testing conditional types and I'd like to write it like this:

```ts
function assert<T extends true>(expectTrue: T extends never ? false : true) {
}

// or even
function assert<T extends true>(expectTrue: T extends true ? true : false) {
}
```

That way I can write:

```ts
type IsNullableType<T> = T extends null | undefined ? true : never;

const result = functionUnderTest(someParam);

assert<IsNullableType<typeof result>>(true);
// or
assert<IsNullableType<typeof result>>(false);
```

So given this:

```ts
assert<IsNullableType<string>>(false);
```

**Expected behavior:**

* The type of the parameter resolves to `false`.
* No error.

**Actual behavior:**

* The type of the parameter resolves to `never`.
* `false` is not assignable to `never`.

**Example where `extends never` works:**

```ts
type IsNonNullableType<T> = IsNullableType<T> extends never ? true : never;
assert<IsNonNullableType<string>>(true); // no compile error, works fine

// this doesn't cause a compile error in playground, but it correctly throws a compile error in the latest version
assert<IsNonNullableType<string | undefined>>(true); // compile error, as expected
```

**Playground Link:** [Link.](http://www.typescriptlang.org/play/#src=function%20assert%3CT%20extends%20true%3E(expectTrue%3A%20T%20extends%20never%20%3F%20false%20%3A%20true)%20%7B%0D%0A%7D%0D%0A%0D%0Atype%20IsNullableType%3CT%3E%20%3D%20T%20extends%20null%20%7C%20undefined%20%3F%20true%20%3A%20never%3B%0D%0Aassert%3CIsNullableType%3Cstring%3E%3E(false)%3B%0D%0A%0D%0Atype%20IsNonNullableType%3CT%3E%20%3D%20IsNullableType%3CT%3E%20extends%20never%20%3F%20true%20%3A%20never%3B%0D%0Aassert%3CIsNonNullableType%3Cstring%3E%3E(true)%3B)

Basically, an `IsNeverType` check does not work: `type IsNeverType<T> = T extends never ? true : never;`

---

Edit: By the way, this is probably a better way to do my check:

```ts
export type IsNullableType<T> = Extract<T, null | undefined> extends never ? false : true;
```

**Comment by dsherret** [comment]
I'm just thinking this might be expected behaviour. So when a type is passed `never` (as `T`) and there's an expression like `T extends ? ... : ...`, then it will resolve that entire `T extends ? ... : ...` to `never`?

Anyway, I resolved my issue by making everything work with `true` or `false`, which is what I probably should have been doing from the start anyway... just couldn't figure out the right way to do it, but `Extract` helped.

**Comment by jack-williams** [comment]
`IsNeverType` needs to be implement like:
```ts
type IsNeverType<T> = [T] extends [never] ? true : never;
```
to stop it being distributive.

**Comment by RyanCavanaugh** [comment]
Seems like a bug to me?
```ts
// Actual instantiation of 'T' shouldn't matter - conditional returns 'number' either way
function assert<T>(expectTrue: T extends never ? number : number) { }
// OK
assert<boolean>(0);
assert<string>(0);
assert<true>(0);
// Error; argument of type '0' not assignable to 'never'
assert<never>(0);
```

**Comment by jack-williams** [comment]
As fair as my understanding goes the conditional doesn't always return `number`, it can also return `never` because it may distribute over nothing.

**Comment by dsherret** [comment]
@RyanCavanaugh good to know! Thanks for looking into this and explaining the behaviour.

@jack-williams thanks a lot for that trick! Very helpful! 😄

**Comment by jcalz** [comment]
Holy cow, this just bit me.  The empty union!

**Comment by pie6k** [comment]
I've created simple helper type

```ts
type NeverAlternative<T, P, N> = [T] extends [never] ? P : N;

// and then
type VariableLengthArray<A, B = never, C = never> =
  //
  NeverAlternative<
    C,
    NeverAlternative<B, [A], [A, B]>,
    NeverAlternative<
      B,
      [A, C],
      [A, B, C]
    >
  >;

type Foo = VariableLengthArray<1,2> // type [1, 2]
type Bar = VariableLengthArray<1,2,3> // type [1, 2, 3]
```

**Comment by loynoir** [comment]
@jack-williams WOW! Awesome! Thanks a lot! 

Is it safe to change return `never` to return `false`?
 
```ts
// type IsNeverType<T> = [T] extends [never] ? true : never;
type IsNeverType<T> = [T] extends [never] ? true : false;
```

**Comment by SyMind** [comment]
> After reading the code I agree with @jack-williams 's assessment. In this case `never` behaves as the _empty union_, so it distributes over the conditional and produces another empty union (which is just `never` again)

Where is the code to do these things? The TypeScript checker code is very complex. I can't find the specific code.
